source/_posts/2017/2017-07-24-how-to-write-custom-fixer-for-php-cs-fixer-24.md

```diff
@@ -0,0 +1,385 @@
+ and [subscribers](/blog/2017/07/17/how-to-write-custom-sniff-for-code-sniffer-3/#2-dispatcher).
+
+
+Yet still, working with tokens is counter intuitive to way we work with the code (class, method, property...), but I'll write about that later.
+
+Now we jump to writing the Fixer class.
+
+
+## 7 Steps to Make an `ExceptionNameFixer` 
+
+"An exception class should have "Exception" suffix."
+
+In last post, we made
+[ExceptionNameSniff](/blog/2017/07/17/how-to-write-custom-sniff-for-code-sniffer-3/#let-s-make-code-exceptionnamesniff-code-together), that will:
+
+- find an "extends" token
+- check if parent class is an Exception
+- find a current class name
+- check if ends "Exception"
+
+Today we'll add one more step:
+
+- **fix the name to end with "Exception"**
+
+
+### 1. Implement an Interface
+
+Create a fixer class and implement a `PhpCsFixer\FixerDefinition\FixerDefinitionInterface` interface. 
+
+It covers 7 required methods, but most of them are easy one-liners:
+
+```php
+use PhpCsFixer\FixerDefinition\FixerDefinitionInterface;
+use PhpCsFixer\Tokenizer\Tokens;
+use SplFileInfo;
+
+final class ExceptionNameFixer implements DefinedFixerInterface
+{
+    # first 5 methods are rutine and descriptive
+
+    public function getName(): string
+    {
+    }
+
+    public function getDefinition(): FixerDefinitionInterface
+    {
+    }        
+
+    public function isRisky(): bool
+    {
+    }        
+
+    public function supports(SplFileInfo $file): bool
+    {
+    }        
+
+    public function getPriority(): int
+    {
+    }        
+    
+    # in last 2 methods, the magic happens :)
+
+    public function isCandidate(Tokens $tokens): bool
+    {
+    }
+
+    public function fix(SplFileInfo $file, Tokens $tokens): void
+    {
+    }
+}
+```
+
+### 2. Easypicks First
+
+I start with implementing first 5 methods, to make
+ 
+```php
+use PhpCsFixer\FixerDefinition\CodeSample;
+use PhpCsFixer\FixerDefinition\FixerDefinition;
+use PhpCsFixer\FixerDefinition\FixerDefinitionInterface;
+use PhpCsFixer\Tokenizer\Tokens;
+use SplFileInfo;
+
+final class ExceptionNameFixer implements DefinedFixerInterface
+{
+    public function getName(): string
+    {
+        return self::class;
+    }
+
+    // this methods return the error message
+    // and it might include a sample code, that would fix it
+    public function getDefinition(): FixerDefinitionInterface
+    {
+        return new FixerDefinition('Exception classes should have suffix "Exception".', [
+            new CodeSample(
+                '<?php
+class SomeClass extends Exception
+{
+}'
+            ),
+        ]);
+    }        
+
+    // if you think this change code in any way, that changes behavior, return "true"
+    // changing a class name is such case
+    public function isRisky(): bool
+    {
+        return true;
+    }        
+
+    // in case you would like to support some specific files
```

Comment:
Only `*.php` are passed via internal iterator.
It is mostly `true`, only difference I found was this: https://github.com/FriendsOfPHP/PHP-CS-Fixer/blob/dc04c34f7f2725a55b84d4b075a7d05586abb495/src/AbstractPsrAutoloadingFixer.php#L55